pkg/term/winconsole/console_windows.go

```diff
@@ -0,0 +1,1132 @@
+}
+
+// mapKeystokeToTerminalString maps the given input event record to string
+func mapKeystokeToTerminalString(keyEvent *KEY_EVENT_RECORD, escapeSequence []byte) string {
+	_, alt, control := getControlKeys(keyEvent.ControlKeyState)
+	if keyEvent.UnicodeChar == 0 {
+		return charSequenceForKeys(keyEvent.VirtualKeyCode, keyEvent.ControlKeyState, escapeSequence)
+	}
+	if control {
+		// TODO(azlinux): Implement following control sequences
+		// <Ctrl>-D  Signals the end of input from the keyboard; also exits current shell.
+		// <Ctrl>-H  Deletes the first character to the left of the cursor. Also called the ERASE key.
+		// <Ctrl>-Q  Restarts printing after it has been stopped with <Ctrl>-s.
+		// <Ctrl>-S  Suspends printing on the screen (does not stop the program).
+		// <Ctrl>-U  Deletes all characters on the current line. Also called the KILL key.
+		// <Ctrl>-E  Quits current command and creates a core
+
+	}
+	// <Alt>+Key generates ESC N Key
+	if !control && alt {
+		return KEY_ESC_N + strings.ToLower(string(keyEvent.UnicodeChar))
+	}
+	return string(keyEvent.UnicodeChar)
+}
+
+// getAvailableInputEvents polls the console for availble events
+// The function does not return until at least one input record has been read.
+func getAvailableInputEvents(fd uintptr) (inputEvents []INPUT_RECORD, err error) {
+	handle := syscall.Handle(fd)
+	if nil != err {
+		return nil, err
+	}
+	for {
+		// Read number of console events available
+		tempBuffer := make([]INPUT_RECORD, MAX_INPUT_BUFFER)
+		nr, err := readConsoleInputKey(uintptr(handle), tempBuffer)
+		if nr == 0 {
+			return nil, err
+		}
+		if 0 < nr {
+			retValue := make([]INPUT_RECORD, nr)
+			for i := 0; i < nr; i++ {
+				retValue[i] = tempBuffer[i]
+			}
+			return retValue, nil
+		}
+	}
+}
+
+// getTranslatedKeyCodes converts the input events into the string of characters
+// The ansi escape sequence are used to map key strokes to the strings
+func getTranslatedKeyCodes(inputEvents []INPUT_RECORD, escapeSequence []byte) string {
+	var buf bytes.Buffer
+	for i := 0; i < len(inputEvents); i++ {
+		input := inputEvents[i]
+		if input.EventType == KEY_EVENT && input.KeyEvent.KeyDown != 0 {
+			keyString := mapKeystokeToTerminalString(&input.KeyEvent, escapeSequence)
+			buf.WriteString(keyString)
+		}
+	}
+	return buf.String()
+}
+
+// ReadChars reads the characters from the given reader
+func (term *WindowsTerminal) ReadChars(fd uintptr, w io.Reader, p []byte) (n int, err error) {
+	n = 0
+	for n < len(p) {
+		select {
+		case b := <-term.inputBuffer:
+			p[n] = b
+			n++
+		default:
+			// Read at least one byte read
+			if n > 0 {
+				return n, nil
+			}
+			inputEvents, _ := getAvailableInputEvents(fd)
+			if inputEvents != nil {
+				if len(inputEvents) == 0 && nil != err {
+					return n, err
+				}
+				if len(inputEvents) != 0 {
+					keyCodes := getTranslatedKeyCodes(inputEvents, term.inputEscapeSequence)
+					for _, b := range []byte(keyCodes) {
+						term.inputBuffer <- b
```

Comment:
i would get rid of the channel and the for loop and simply do `n += copy(p[n:], keyCodes)`